In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [ ]:
df = pd.read_csv('/content/healthcare-dataset-stroke-data (1).csv')
df

gender   age  hypertension  heart_disease ever_married      work_type   
0       Male  67.0             0              1          Yes        Private  \
1     Female  61.0             0              0          Yes  Self-employed   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
...      ...   ...           ...            ...          ...            ...   
5104  Female  80.0             1              0          Yes        Private   
5105  Female  81.0             0              0          Yes  Self-employed   
5106  Female  35.0             0              0          Yes  Self-employed   
5107    Male  51.0             0              0          Yes        Private   
5108  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0             Urban             228.69  36.6  formerly smoked       1  
1             Rural             202.21   NaN     never smoked       1  
2             Rural             105.92  32.5     never smoked       1  
3             Urban             171.23  34.4           smokes       1  
4             Rural             174.12  24.0     never smoked       1  
...             ...                ...   ...              ...     ...  
5104          Urban              83.75   NaN     never smoked       0  
5105          Urban             125.20  40.0     never smoked       0  
5106          Rural              82.99  30.6     never smoked       0  
5107          Rural             166.29  25.6  formerly smoked       0  
5108          Urban              85.28  26.2          Unknown       0  

[5109 rows x 11 columns]

In [ ]:
# Convert categorical variables to numerical using LabelEncoder
categorical_cols = ['gender', 'ever_married', 'work_type', 'Residence_type', 'smoking_status']
label_encoders = {}
for col in categorical_cols:
    label_encoders[col] = LabelEncoder()
    df[col] = label_encoders[col].fit_transform(df[col])

df

gender   age  hypertension  heart_disease  ever_married  work_type   
0          1  67.0             0              1             1          2  \
1          0  61.0             0              0             1          3   
2          1  80.0             0              1             1          2   
3          0  49.0             0              0             1          2   
4          0  79.0             1              0             1          3   
...      ...   ...           ...            ...           ...        ...   
5104       0  80.0             1              0             1          2   
5105       0  81.0             0              0             1          3   
5106       0  35.0             0              0             1          3   
5107       1  51.0             0              0             1          2   
5108       0  44.0             0              0             1          0   

      Residence_type  avg_glucose_level   bmi  smoking_status  stroke  
0                  1             228.69  36.6               1       1  
1                  0             202.21   NaN               2       1  
2                  0             105.92  32.5               2       1  
3                  1             171.23  34.4               3       1  
4                  0             174.12  24.0               2       1  
...              ...                ...   ...             ...     ...  
5104               1              83.75   NaN               2       0  
5105               1             125.20  40.0               2       0  
5106               0              82.99  30.6               2       0  
5107               0             166.29  25.6               1       0  
5108               1              85.28  26.2               0       0  

[5109 rows x 11 columns]

In [ ]:
df.isnull()

gender    age  hypertension  heart_disease  ever_married  work_type   
0      False  False         False          False         False      False  \
1      False  False         False          False         False      False   
2      False  False         False          False         False      False   
3      False  False         False          False         False      False   
4      False  False         False          False         False      False   
...      ...    ...           ...            ...           ...        ...   
5104   False  False         False          False         False      False   
5105   False  False         False          False         False      False   
5106   False  False         False          False         False      False   
5107   False  False         False          False         False      False   
5108   False  False         False          False         False      False   

      Residence_type  avg_glucose_level    bmi  smoking_status  stroke  
0              False              False  False           False   False  
1              False              False   True           False   False  
2              False              False  False           False   False  
3              False              False  False           False   False  
4              False              False  False           False   False  
...              ...                ...    ...             ...     ...  
5104           False              False   True           False   False  
5105           False              False  False           False   False  
5106           False              False  False           False   False  
5107           False              False  False           False   False  
5108           False              False  False           False   False  

[5109 rows x 11 columns]

In [ ]:
#replace these null values with mean value ( avg ) of the column

#mean-value of bmi
mean = np.mean(df['bmi'])
mean

#use fillna to replace null-values

df['bmi'].fillna(mean,inplace=True)
df

gender   age  hypertension  heart_disease  ever_married  work_type   
0          1  67.0             0              1             1          2  \
1          0  61.0             0              0             1          3   
2          1  80.0             0              1             1          2   
3          0  49.0             0              0             1          2   
4          0  79.0             1              0             1          3   
...      ...   ...           ...            ...           ...        ...   
5104       0  80.0             1              0             1          2   
5105       0  81.0             0              0             1          3   
5106       0  35.0             0              0             1          3   
5107       1  51.0             0              0             1          2   
5108       0  44.0             0              0             1          0   

      Residence_type  avg_glucose_level       bmi  smoking_status  stroke  
0                  1             228.69  36.60000               1       1  
1                  0             202.21  28.89456               2       1  
2                  0             105.92  32.50000               2       1  
3                  1             171.23  34.40000               3       1  
4                  0             174.12  24.00000               2       1  
...              ...                ...       ...             ...     ...  
5104               1              83.75  28.89456               2       0  
5105               1             125.20  40.00000               2       0  
5106               0              82.99  30.60000               2       0  
5107               0             166.29  25.60000               1       0  
5108               1              85.28  26.20000               0       0  

[5109 rows x 11 columns]

In [ ]:
##Normalization = 0-1 Range

from sklearn.preprocessing import MinMaxScaler

#make age,bmi,glucose-levels in 0-1 range

df[['gender','age','work_type','avg_glucose_level','bmi','smoking_status']] = MinMaxScaler().fit_transform(df[['gender','age','work_type','avg_glucose_level','bmi','smoking_status']])
df

gender       age  hypertension  heart_disease  ever_married  work_type   
0        0.5  0.816895             0              1             1       0.50  \
1        0.0  0.743652             0              0             1       0.75   
2        0.5  0.975586             0              1             1       0.50   
3        0.0  0.597168             0              0             1       0.50   
4        0.0  0.963379             1              0             1       0.75   
...      ...       ...           ...            ...           ...        ...   
5104     0.0  0.975586             1              0             1       0.50   
5105     0.0  0.987793             0              0             1       0.75   
5106     0.0  0.426270             0              0             1       0.75   
5107     0.5  0.621582             0              0             1       0.50   
5108     0.0  0.536133             0              0             1       0.00   

      Residence_type  avg_glucose_level       bmi  smoking_status  stroke  
0                  1           0.801265  0.301260        0.333333       1  
1                  0           0.679023  0.212996        0.666667       1  
2                  0           0.234512  0.254296        0.666667       1  
3                  1           0.536008  0.276060        1.000000       1  
4                  0           0.549349  0.156930        0.666667       1  
...              ...                ...       ...             ...     ...  
5104               1           0.132167  0.212996        0.666667       0  
5105               1           0.323516  0.340206        0.666667       0  
5106               0           0.128658  0.232532        0.666667       0  
5107               0           0.513203  0.175258        0.333333       0  
5108               1           0.139230  0.182131        0.000000       0  

[5109 rows x 11 columns]

In [ ]:
# Split the data into features (X) and target variable (y)
X = df.drop('stroke', axis=1)
y = df['stroke']

# Unbalanced Data

In [ ]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Decision Tree Classifier

In [ ]:
# Create a decision tree classifier
clf = DecisionTreeClassifier()

In [ ]:
# Train the classifier on the training data
clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
# Make predictions on the test data
y_pred = clf.predict(X_test)


In [ ]:
# Evaluate the accuracy of the classifier
accuracy = metrics.accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9093281148075668


# Support Vector Machine (SVM)

In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [ ]:
# Split the data into features (X) and target variable (y)
X = df.drop('stroke', axis=1)
y = df['stroke']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [ ]:
# Create an SVM classifier
clf = SVC()

In [ ]:

# Train the classifier on the training data
clf.fit(X_train, y_train)

SVC()

In [ ]:
# Make predictions on the test data
y_pred = clf.predict(X_test)

# Calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9419439008480104


In [ ]:
# Calculate performance metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)

# Print the performance metrics
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_mat)

Accuracy: 0.7757201646090535
Precision: 0.7593548387096775
Recall: 0.8072702331961591
F1 Score: 0.7825797872340426
Confusion Matrix:
 [[1085  373]
 [ 281 1177]]


Accuracy achieved using decision tree: 0.9060665362035225

Accuracy achieved using SVM: 0.9419439008480104

# Balancing the Data

In [ ]:
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

In [ ]:
# Perform SMOTE to balance the data
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

In [ ]:
# Display the count of each class after balancing
class_counts = pd.Series(y_resampled).value_counts()
print("Class Counts after balancing:\n", class_counts)

Class Counts after balancing:
 stroke
1    4860
0    4860
Name: count, dtype: int64


In [ ]:
# Split the resampled data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, random_state=42)

# Decision tree after balancing data

In [ ]:
# Create a decision tree classifier
clf = DecisionTreeClassifier()

In [ ]:
# Train the classifier on the training data
clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
# Make predictions on the test data
y_pred = clf.predict(X_test)


In [ ]:
# Evaluate the accuracy of the classifier
accuracy = metrics.accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8885459533607681


# SVM after balancing data

In [ ]:
# Create an SVM classifier
clf = SVC()

In [ ]:
# Train the classifier on the training data
clf.fit(X_train, y_train)

SVC()

In [ ]:
# Make predictions on the test data
y_pred = clf.predict(X_test)

In [ ]:
# Calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8240740740740741


# Accuracy achieved after balancing and using Decision Tree: 0.8878600823045267

# Accuracy achieved after balancing and using SVM: 0.8240740740740741

# Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
# Perform SMOTE to balance the data
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

In [ ]:
# Split the resampled data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, random_state=42)

Decision tree

In [ ]:
# Create a Decision Tree classifier
dt_clf = DecisionTreeClassifier()

In [ ]:
# Train the Decision Tree classifier on the training data
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
# Make predictions on the test data using the Decision Tree classifier
dt_y_pred = dt_clf.predict(X_test)


In [ ]:
# Generate the confusion matrix for Decision Tree classifier
dt_cm = confusion_matrix(y_test, dt_y_pred)
print("Decision Tree Confusion Matrix:\n", dt_cm)

Decision Tree Confusion Matrix:
 [[1299  159]
 [ 142 1316]]


Support Vector machine(SVM)

In [ ]:
# Create an SVM classifier
svm_clf = SVC()

In [ ]:
# Train the SVM classifier on the training data
svm_clf.fit(X_train, y_train)

SVC()

In [ ]:
# Make predictions on the test data using the SVM classifier
svm_y_pred = svm_clf.predict(X_test)

In [ ]:
# Generate the confusion matrix for SVM classifier
svm_cm = confusion_matrix(y_test, svm_y_pred)
print("SVM Confusion Matrix:\n", svm_cm)

SVM Confusion Matrix:
 [[1104  354]
 [ 159 1299]]


# ensemble learning

# Random Forest in balanced data

In [ ]:
# Perform SMOTE to balance the data
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

In [ ]:
# Split the resampled data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, random_state=42)

In [ ]:
# Create a Random Forest classifier
clf = RandomForestClassifier()

In [ ]:
# Train the classifier on the training data
clf.fit(X_train, y_train)

RandomForestClassifier()

In [ ]:
# Make predictions on the test data
y_pred = clf.predict(X_test)

In [ ]:
# Evaluate the accuracy of the classifier
accuracy = metrics.accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9502743484224966


# Adaboost Classifier

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
# Create an AdaBoost classifier
adaboost_clf = AdaBoostClassifier()

In [ ]:
# Train the AdaBoost classifier on the training data
adaboost_clf.fit(X_train, y_train)

AdaBoostClassifier()

In [ ]:
# Make predictions on the test data
y_pred = adaboost_clf.predict(X_test)

In [ ]:
# Calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8590534979423868


# Conclusion

Accuracy achieved using random forest: 0.9502743484224966

Accuracy achieved using AdaBoost: 0.8590534979423868

Among all the models, the Random Forest Classifier achieved the highest accuracy of 95% on the balanced data.